EIPS/eip-4907.md

```diff
@@ -0,0 +1,185 @@
+    function setUser(uint256 tokenId, address user, uint64 expires) external;
+
+    /// @notice Get the user address of an NFT
+    /// @dev The zero address indicates that there is no user or the user is expired
+    /// @param tokenId The NFT to get the user address for
+    /// @return The user address for this NFT
+    function userOf(uint256 tokenId) external view returns(address);
+
+    /// @notice Get the user expires of an NFT
+    /// @dev The zero value indicates that there is no user
+    /// @param tokenId The NFT to get the user expires for
+    /// @return The user expires for this NFT
+    function userExpires(uint256 tokenId) external view returns(uint256);
+}
+```
+
+The `userOf(uint256 tokenId)` function MAY be implemented as `pure` or `view`.
+
+The `userExpires(uint256 tokenId)` function MAY be implemented as `pure` or `view`.
+
+The `setUser(uint256 tokenId, address user, uint64 expires)` function MAY be implemented as `public` or `external`.
+
+The `UpdateUser` event MUST be emitted when a user address is changed or the user expires is changed.
+
+The `supportsInterface` method MUST return `true` when called with `0xad092b5c`.
+
+## Rationale
+
+Many developers are trying to develop based on the NFT utility, and some of them have added roles already,  but there are some key problems need to be solved. The advantages of this standard are below.
+
+### Clear Permissions Management
+
+Usage rights are part of ownership, so **owner** can modify **user** at any time, while **user** is only granted some specific permissions, such as **user** usually does not have permission to make permanent changes to NFT's Metadata.
+
+NFTs may be used in multiple applications, and adding the user role to NFTs makes it easier for the application to make special grants of rights.
+
+### Simple On-chain Time Management
+
+Most NFTs do not take into account the expiration time even though the role of the user is added, resulting in the need for the owner to manually submit on-chain transaction to cancel the user rights, which does not allow accurate on-chain management of the use time and will waste gas.
+
+The usage right often corresponds to a specific time, such as deploying scenes on land, renting game props,  etc. Therefore, it can reduce the on-chain transactions and save gas with **expires**.
+
+### Easy Third-Party Integration
+
+The standard makes it easier for third-party protocols to manage NFT usage rights without permission from the NFT issuer or the NFT application.
+
+## Backwards Compatibility
+
+As mentioned in the specifications section, this standard can be fully ERC721 compatible by adding an extension function set.
+
+In addition, new functions introduced in this standard have many similarities with the existing functions in ERC721. This allows developers to easily adopt the standard quickly.
+
+## Test Cases
+
+Test cases for EIP4907 are at https://github.com/emojidao/EIP4907 .
```

Comment:
Please either inline the test cases in this document, or include them in the `assets/eip-4907` directory.